### PR TITLE
Refactor/dutch x

### PR DESCRIPTION
**Pull request by Velenir**
WIP

more abstraction and modularity for:
+ web3
+ Tokens
+ DutchX
+ an overarching contracts/account api in `api/index.js`

### CODE REVIEW

File: src/api/DutchX.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ async function init(): Promise<DutchExchange> {
   }
 
   const getClosingPrice = async (pair: TokenPair, index?: number) => {
-    if (index === undefined) index = await getAuctionIndex(pair)
+    if (index === undefined) index = await getAuctionIndex(pair) - 1
```
Comment (W3stside) [inline src/api/DutchX.ts]: this is what needs to be taken out yeah?

File: src/api/index.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+  // account would normally be taken from redux state and passed inside an action
+  // but just in case
+  if (!account) account = await web3.getCurrentAccount()
+
+  return Tokens.getTokenBalance('ETH', account)
+}
+
+// TODO: remove ['ETH', 'GNO'] default, use actions for this
+export const getTokenBalances = async (codeList: TokenCode[] = ['ETH', 'GNO'], account?: Account) => {
```
Comment (W3stside) [inline src/api/index.ts]: why is this called codeList and not tokenList?

Comment (Velenir) [reply src/api/index.ts]: don't know, will change to `tokenList`

File: src/api/DutchX.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    return index.toNumber()
+  }
+
+  const getClosingPrice = async (pair: TokenPair, index?: number) => {
+    if (index === undefined) index = await getAuctionIndex(pair) - 1
+
+    const [num, den] = getExchange(pair).closingPrices(index)
+
+    return num.div(den).toString()
```
Comment (W3stside) [inline src/api/DutchX.ts]: Not super positive with the changes that will come - but I think we should be careful of this return of `num/den` as we may need to divide the inverse when calculating the other side of Token Exchanges _e.g `ETH/GNO` && `GNO/ETH`_

File: src/api/DutchX.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+    return vol.toString()
+  }
+
+  const getSellerBalances = async (pair: TokenPair, index?: number, account?: Account) => {
+    [index, account] = await fillDefaultIndexAndAccount(pair, index, account)
+
+    const balance = await getExchange(pair).sellerBalances(index, account)
+
+    return await balance.toString()
```
Comment (W3stside) [inline src/api/DutchX.ts]: why `await` here?

File: src/api/web3Provider.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  const resetProvider = () => setProvider(getProvider())
+
+  return {
+    getCurrentAccount,
+    getAccounts,
+    getETHBalance,
+    getNetwork,
+    isConnected,
+    get currentProvider() {
```
Comment (W3stside) [inline src/api/web3Provider.ts:56]: This is new to me - are you declaring the function here _and_ calling the getter?

Comment (Velenir) [reply src/api/web3Provider.ts:56]: declaring the getter, same as
```
class CL {
  get prop () {
    return this.prop
  }
}
```

File: src/api/DutchX.ts (hunk 4 above)
Comment (Velenir) [reply src/api/DutchX.ts]: nice spotting skills. will remove

File: src/api/contracts.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  'DutchExchange',
+  'DutchExchangeFactory',
+  'DutchExchangeETHGNO',
+  'DutchExchangeGNOETH',
+  'Token',
+  'TokenETH',
+  'TokenGNO',
+]
+
+const Contracts = contractNames.map(name => TruffleContract(require(`../../build/contracts/${name}.json`)))
```
Comment (W3stside) [inline src/api/contracts.ts]: Could you rename to `ContractJSON` or `ContractArtifacts` ? 😅

Comment (Velenir) [reply src/api/contracts.ts]: These are not contract artifacts, they are contract classes, if you will. Since each is `TruffleContract(artifact)`

Comment (W3stside) [reply src/api/contracts.ts]: oh I didn't see the TruffleContract before, sneaky. Nvm then

File: src/api/utils/index.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  window.addEventListener('load', function loadHandler(event) {
+    window.removeEventListener('load', loadHandler, false)
+    return accept(event)
+  }, false)
+})
+
+
+export const promisify = (func: Function, context: object, ...defArgs: any[]) =>
```
Comment (W3stside) [inline src/api/utils/index.ts:17]: I hate this function 😂 😂 😂 😂

Comment (Velenir) [reply src/api/utils/index.ts:17]: I'm open to any changes. This one is actually a slightly modified `util.promisify` from node, but we can't use node system libraries in the browser(

File: src/components/BalanceButton/index.tsx [hunk 8]
```diff
@@ -1,6 +1,6 @@
 /* eslint no-console:0 */
```
Comment (W3stside) [inline src/components/BalanceButton/index.tsx]: can you delete this component, not used

Comment (W3stside) [inline src/components/BalanceButton/index.tsx]: can you delete this component, not used

File: src/api/utils/index.ts (hunk 7 above)
Comment (W3stside) [reply src/api/utils/index.ts:17]: its fine for now

**Review by W3stside — APPROVED** [review]
From testing and going through code it looks good, merge

**Pull request merged**